router/types/types.go

```diff
@@ -1,24 +1,366 @@
 package router
 
 import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/hex"
 	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"strings"
 	"time"
 )
 
 // Certificate describes a TLS certificate for one or more routes
 type Certificate struct {
-	// ID is the unique ID of this Certificate
-	ID string `json:"id,omitempty"`
 	// Routes contains the IDs of routes assigned to this cert
-	Routes []string `json:"routes,omitempty"`
-	// TLSCert is the optional TLS public certificate. It is only used for HTTP routes.
-	Cert string `json:"cert,omitempty"`
-	// TLSCert is the optional TLS private key. It is only used for HTTP routes.
-	Key string `json:"key,omitempty"`
+	Routes []string
+	// Chain is the DER-encoded TLS certificate chain.
+	Chain [][]byte
+	// Key is the DER-encoded TLS private key.
+	Key []byte
+	// NoStrict is whether to skip performing strict checks on the
+	// certificate
+	NoStrict bool
 	// CreatedAt is the time this cert was created.
-	CreatedAt time.Time `json:"created_at,omitempty"`
+	CreatedAt time.Time
 	// UpdatedAt is the time this cert was last updated.
-	UpdatedAt time.Time `json:"updated_at,omitempty"`
+	UpdatedAt time.Time
+
+	// redactJSONKey when set will appear as the "key" field when the
+	// certificate is encoded as JSON
+	redactJSONKey string
+}
+
+// ID returns the unique ID of this Certificate
+func (c *Certificate) ID() string {
+	digest := sha256.Sum256(bytes.Join(c.Chain, []byte{}))
+	return hex.EncodeToString(digest[:])
+}
+
+func (c *Certificate) SetRedactJSONKey(s string) {
+	c.redactJSONKey = s
+}
+
+type certificateJSON struct {
+	ID             string    `json:"id,omitempty"`
+	Routes         []string  `json:"routes,omitempty"`
+	Chain          string    `json:"chain,omitempty"`
+	Key            string    `json:"key,omitempty"`
+	CreatedAt      time.Time `json:"created_at,omitempty"`
+	UpdatedAt      time.Time `json:"updated_at,omitempty"`
+	DeprecatedCert string    `json:"cert,omitempty"`
+}
+
+func (c *Certificate) MarshalJSON() ([]byte, error) {
+	key := c.redactJSONKey
+	if key == "" {
+		key = c.KeyPEM()
```

Comment:
The key field should never be exposed via the API (except the redacted version that shows up in the logs to indicate that the field was set).

Reply:
The reason the key was in the JSON was because that's how the router would get it (specifically from a controller route event that contains the JSON encoded route + certificate).

I've replaced the key with the keyID in the JSON and added a separate `GetPrivateKey` gRPC method that the router uses to get the key (which I guess we can ultimately lock down further so only the router can call that method).

See e1905c2b.